src/transformStateWithClones.js

```diff
@@ -7,7 +7,39 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateCopy = { ...state };
+  const stateHistory = [];
+
+  for (const action of actions) {
+    const { type, extraData, keysToRemove } = action;
+
+    switch (type) {
+      case 'addProperties':
+        Object.assign(stateCopy, extraData);
+        break;
+
+      case 'removeProperties':
+        keysToRemove.forEach(function (key) {
+          delete stateCopy[key];
+        });
+        break;
+
+      case 'clear':
+        for (const key in stateCopy) {
+          if (stateCopy.hasOwnProperty(key)) {
+            delete stateCopy[key];
+          }
+        }
+        break;
+
+      default:
+        break;
+    }
+
+    stateHistory.push({ ...stateCopy });
```

Comment:
[DONT REPEAT YOURSELF] - It is better to push a copy of the object to the array at the end of the each loop cycle, but outside of the `switch` block, after current action is processed. However, since you are currently mutating the same `stateCopy` throughout the loop, this would not work correctly until you address the issue of creating a new copy inside the loop for each action.